src/transformers/generation/utils.py

```diff
@@ -3404,184 +3439,241 @@ def _beam_search(
+                topk_indices = torch.multinomial(nn.functional.softmax(log_probs, dim=-1), num_samples=beams_to_keep)
+                topk_log_probs = torch.gather(input=log_probs, dim=1, index=topk_indices)
             else:
-                next_token_scores, next_tokens = torch.topk(
-                    next_token_scores, n_tokens_to_keep, dim=1, largest=True, sorted=True
-                )
-
-            next_indices = torch.div(next_tokens, vocab_size, rounding_mode="floor")
-            next_tokens = next_tokens % vocab_size
+                topk_log_probs, topk_indices = torch.topk(log_probs, k=beams_to_keep)
+
+            # Gather K top beams, recover the beam index by floor division and token id by modulo division
+            topk_current_beam_indices = topk_indices // vocab_size
+            topk_running_beam_indices = gather_beams(running_beam_indices, topk_current_beam_indices)
+            topk_running_sequences = gather_beams(running_sequences, topk_current_beam_indices)
+            topk_ids = topk_indices % vocab_size
+
+            # Update sequences for the K top-k new sequences.
+            topk_running_sequences = torch.cat([topk_running_sequences, topk_ids[:, :, None]], dim=-1)
+
+            # we want to store the beam indices with batch information -> real beam index = beam index % num beams
+            batch_offset = torch.arange(batch_size, device=input_ids.device).view(-1, 1) * num_beams
+            batch_modified_indices = topk_current_beam_indices + batch_offset
+            topk_running_beam_indices = torch.cat(
+                (topk_running_beam_indices, batch_modified_indices[:, :, None]), dim=-1
+            )
 
-            # stateless
-            beam_outputs = beam_scorer.process(
-                input_ids,
-                next_token_scores,
-                next_tokens,
-                next_indices,
-                pad_token_id=pad_token_id,
-                eos_token_id=eos_token_id,
-                beam_indices=beam_indices,
-                decoder_prompt_len=decoder_prompt_len,
+            # d. Check which sequences have ended
+            next_token_hits_stopping_criteria = stopping_criteria(flatten_beam_dim(topk_running_sequences), all_scores)
+            next_token_hits_stopping_criteria = unflatten_beam_dim(
+                next_token_hits_stopping_criteria, batch_size, beams_to_keep
             )
 
-            beam_scores = beam_outputs["next_beam_scores"]
-            beam_next_tokens = beam_outputs["next_beam_tokens"]
-            beam_idx = beam_outputs["next_beam_indices"]
+            # Only the top `num_beam` sequences can be considered for the final returned sequences
+            top_num_beam_mask = torch.cat(
+                (
+                    torch.ones((num_beams), dtype=torch.bool),
+                    torch.zeros((beams_to_keep - num_beams), dtype=torch.bool),
+                ),
+                dim=0,
```

Comment:
this is a bit costly, we could have a tensor of size num_beams + to keep + x that we just write to / slice

Reply:
Moved out of the loop: it is a constant, doesn't need to be computed once per iteration